sklearn/preprocessing/tests/test_target_encoder.py

```diff
@@ -70,26 +70,26 @@ def test_encoding(categories, unknown_value, global_random_seed, smooth, target_
 
     if categories == "auto":
         X_train = X_train_array
-    else:
-        X_train = categories[0][X_train_array]
-
-    if categories == "auto":
         X_test = X_test_array
     else:
+        X_train = categories[0][X_train_array]
         X_test = categories[0][X_test_array]
+
     X_test = np.concatenate((X_test, [[unknown_value]]))
 
     rng = np.random.RandomState(global_random_seed)
 
+    n_splits = 3
+    random_state = 0
```

Comment:
I think we should be using the `rng` which takes from `global_random_seed` instead?

Reply:
Glad you raised this because it uncovers interesting behaviour. The tests fail if we use `rng` in the *Kfold and also pass it to `TargetEncoder`. I think this is because we are calling shuffle on this `RandomState` object more than once, which gives us different results (I think this is intended behaviour), see:

```
In [1]: import numpy as np

In [2]: rng = np.random.RandomState(42)

In [3]: a = np.arange(10)

In [4]: a
Out[4]: array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [5]: rng.shuffle(a)

In [6]: a
Out[6]: array([8, 1, 5, 0, 7, 2, 9, 4, 3, 6])

In [7]: a = np.arange(10)

In [8]: a
Out[8]: array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [9]: rng.shuffle(a)

In [10]: a
Out[10]: array([0, 1, 8, 5, 3, 4, 7, 9, 6, 2])
```

Generating a new `RandomState` object (with the same seed) and then calling shuffle on `a` will give the same results as "Out[6]". Thus when we use an `int` a new `RandomState` object is created each time and the test pass.

Not sure if we need/can to do anything about this? Potentially just document?

Reply:
I'm not sure if I understand. If we need to get a shuffle and re-use it, we should be doing that and passing it along, and if that's not the case, I don't see why it would make the tests fail.

Reply:
I am not sure I follow you either. We shuffle separately in `for train_idx, test_idx in cv.split(X_train_array, y_train):` and inside of `TargetEncoder.fit_transform`. We can't get a shuffle and re-use it (if re-use means reusing the shuffled indices).

But for clarity, the symptoms are at least: the tests fail when I use `rng` and pass when I use `random_state`. The shuffled indices (e.g., taking the continuous case, `KFold`) from here:

https://github.com/scikit-learn/scikit-learn/blob/594475a1726ba02a943e5f2218566db28e184e88/sklearn/model_selection/_split.py#L482

are different when done in the test at the line `for train_idx, test_idx in cv.split(X_train_array, y_train):` and when done inside of `TargetEncoder.fit_transform`.

I checked (with `id()`) and the `rng` object used in and in the line `for train_idx, test_idx in cv.split(X_train_array, y_train):` and in `TargetEncoder.fit_transform`

Reply:
Ah im so silly. I will just create a new `rng` and pass it to `TargetEncoder`

Reply:
Okay the only way to make it work using `rng` is to create a new instance before it is passed to a *KFold or `TargetEncoder`. This is because `rng.shufffle` will give different results if the `rng` instance has been used previously.

Current changes work, (note ran CIs just to show that using the same `rng` did not work, see: https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=57690&view=logs&j=dde5042c-7464-5d47-9507-31bdd2ee0a3a&t=4bd2dad8-62b3-5bf9-08a5-a9880c530c94&l=1616)

@adrinjalali let me know if these changes are okay?

Reply:
Okay, had to use use a new variable name for `rng` because if I used the same, it would not create a new instance (i.e. `id()` would be the same sometimes).

Summary:
* Creating ONE `rng` and passing it to `random_state` in `StratifiedKFold`/`KFold` and `TargetEncoder` did not work: if the `rng` instance has been used before `shuffle` would produce different results (CI fails in commit [3b815a1](https://github.com/scikit-learn/scikit-learn/pull/26958/commits/3b815a113ad1e44191876ddefd5b0835d98aece5)) 
* Creating a new `RandomState` instance and naming it `rng` before passing to `StratifiedKFold`/`KFold` and `TargetEncoder` did not work - it seems that sometimes no new instance was created (i.e., `id()` was the same) (CIs failed in: [3f741c3](https://github.com/scikit-learn/scikit-learn/pull/26958/commits/3f741c30ec87ff3e1967cac63f1bc71e12163eb9))
* When creating a new `RandomState` instance (with the same seed), always name it something different from what it was name before works: [86a5f5b](https://github.com/scikit-learn/scikit-learn/pull/26958/commits/86a5f5bdf6c8bde10584261f64c1a2779c541fab) passes

Reply:
Ah I see this is a known issue: #14042

I think it would be best to go back to using an int.

Reply:
I think the current tests refactor is better than `main` and good enough to be merge as is. I think there are ways to get `global_random_state` to work, but that can be done as a follow up.

Reply:
Just realised that a reasonably neat way is to `set_state` each time before it passed to an estimator, done in: [f6ef1d2](https://github.com/scikit-learn/scikit-learn/pull/26958/commits/f6ef1d2be87c8a1bc2a16e37154dffc9184b5de2)
However, happy to revert if you prefer @thomasjpfan @adrinjalali

Reply:
I am okay with `get_state` + `get_state` as long as there is a comment above `get_state` explaining why it is necessary.

Reply:
Done